db/database.go

```diff
@@ -1071,151 +1074,189 @@ func (db *Database) UpdateAllDocChannels() (int, error) {
 
 	base.Infof(base.KeyAll, "Recomputing document channels...")
 
-	results, err := db.QueryResync()
-	if err != nil {
-		return 0, err
-	}
-
 	// We are about to alter documents without updating their sequence numbers, which would
 	// really confuse the changeCache, so turn it off until we're done:
 	db.changeCache.EnableChannelIndexing(false)
 	defer db.changeCache.EnableChannelIndexing(true)
 
-	err = db.changeCache.Clear()
+	err := db.changeCache.Clear()
 	if err != nil {
 		return 0, err
 	}
 
 	base.Infof(base.KeyAll, "Re-running sync function on all documents...")
-	changeCount := 0
-	docCount := 0
-
-	var importRow QueryIdRow
-	for results.Next(&importRow) {
-		docid := importRow.Id
-		key := realDocID(docid)
-
-		docCount++
-		documentUpdateFunc := func(doc *Document) (updatedDoc *Document, shouldUpdate bool, updatedExpiry *uint32, err error) {
-			imported := false
-			if !doc.HasValidSyncData() {
-				// This is a document not known to the sync gateway. Ignore it:
-				return nil, false, nil, base.ErrUpdateCancel
-			} else {
-				base.Debugf(base.KeyCRUD, "\tRe-syncing document %q", base.UD(docid))
-			}
 
-			// Run the sync fn over each current/leaf revision, in case there are conflicts:
-			changed := 0
-			doc.History.forEachLeaf(func(rev *RevInfo) {
-				bodyBytes, _, err := db.get1xRevFromDoc(doc, rev.ID, false)
-				if err != nil {
-					base.Warnf("Error getting rev from doc %s/%s %s", base.UD(docid), rev.ID, err)
-				}
-				var body Body
-				if err := body.Unmarshal(bodyBytes); err != nil {
-					base.Warnf("Error unmarshalling body %s/%s for sync function %s", base.UD(docid), rev.ID, err)
-					return
+	queryLimit := db.Options.ResyncQueryLimit
+	startSeq := uint64(0)
+	endSeq, err := db.sequences.lastSequence()
+	if err != nil {
+		return 0, err
+	}
+
+	// Reset these values now that a new run has begun
+	db.ResyncManager.ResetStatus()
+
+	docsChanged := 0
+	docsProcessed := 0
+
+	// In the event of an early exit we would like to ensure these values are up to date which they wouldn't be if they
+	// were unable to reach the end of the batch iteration.
+	defer func() {
+		db.ResyncManager.UpdateProcessedChanged(docsProcessed, docsChanged)
+	}()
+
+	for {
+		results, err := db.QueryResync(queryLimit, startSeq, endSeq)
+		if err != nil {
+			return 0, err
+		}
+
+		queryRowCount := 0
+		highSeq := uint64(0)
+
+		var importRow QueryIdRow
+		for results.Next(&importRow) {
+			if db.ResyncManager.ResyncTerminator.CompareAndSwap(true, false) {
+				base.Infof(base.KeyAll, "Re-running of Sync Function was stopped before the operation could be "+
```

Comment:
Just 'Resync' is better than 'Re-running of Sync Function' here I think, as it sounds like an individual execution of the Sync Function is what was stopped here.

Reply:
Also, if the terminator is set by 'stop' after this iteration completes (but before resync ends), it's going to cause the *next* resync to fail.  Need to ensure the terminator is reset on resync start as well as the stats.